xray/src/inpaint.rs

```diff
@@ -0,0 +1,264 @@
+                })
+            })
+            .transpose()
+    }
+
+    fn stitched_image(&self) -> ImageResult<Option<RgbaImage>> {
+        if let Some(current) = self.image_from(None)? {
+            let w = current.width() / 2;
+            let h = current.height() / 2;
+            let mut image = RgbaImage::from_pixel(4 * w, 4 * h, Rgba::from(TRANSPARENT.to_u8()));
+            // TODO(feuerste): Once we use image >= 0.23, add the ? operator
+            image.copy_from(&current, w, h);
+            let mut copy_subimage = |direction: Direction,
+                                     from_x: u32,
+                                     from_y: u32,
+                                     width: u32,
+                                     height: u32,
+                                     to_x: u32,
+                                     to_y: u32|
+             -> ImageResult<()> {
+                if let Some(neighbor) = self.image_from(direction)? {
+                    // TODO(feuerste): Once we use image >= 0.23, add the ? operator
+                    image.copy_from(&neighbor.view(from_x, from_y, width, height), to_x, to_y);
+                }
+                Ok(())
+            };
+            copy_subimage(Direction::TopLeft, w, h, w, h, 0, 0)?;
+            copy_subimage(Direction::Top, 0, h, 2 * w, h, w, 0)?;
+            copy_subimage(Direction::TopRight, 0, h, w, h, 3 * w, 0)?;
+            copy_subimage(Direction::Right, 0, 0, w, 2 * h, 3 * w, h)?;
+            copy_subimage(Direction::BottomRight, 0, 0, w, h, 3 * w, 3 * h)?;
+            copy_subimage(Direction::Bottom, 0, 0, 2 * w, h, w, 3 * h)?;
+            copy_subimage(Direction::BottomLeft, w, 0, w, h, 0, 3 * h)?;
+            copy_subimage(Direction::Left, w, 0, w, 2 * h, 0, h)?;
+            Ok(Some(image))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn create_inpaint_image(&self, inpaint_distance_px: u8) -> ImageResult<()> {
+        if let Some(mut inpaint_image) = self.stitched_image()? {
+            inpaint_image = inpaint(inpaint_image, inpaint_distance_px);
+            let inpaint_image_path = self.get_inpaint_image_path();
+            inpaint_image.save(inpaint_image_path)?;
+        }
+        Ok(())
+    }
+
+    fn interpolate_inpaint_image_with_right(&self) -> ImageResult<()> {
+        if let (Some((mut current, current_path)), Some((mut right, right_path))) = (
+            self.inpaint_image_and_path_from(None)?,
+            self.inpaint_image_and_path_from(Direction::Right)?,
+        ) {
+            let (width, height) = (current.width() / 2, current.height());
+            interpolate_subimages(
+                &mut right.sub_image(0, 0, width, height),
+                &mut current.sub_image(width, 0, width, height),
+                |i, _| i as f32 / (width - 1) as f32,
+            );
+            current.save(current_path)?;
+            right.save(right_path)?;
+        }
+        Ok(())
+    }
```

Comment:
I think we could replace this and the next function by one, more generic, function:
```
    fn interpolate_inpaint_image_with_neighbor(&self, direction::Direction) -> ImageResult<()> {
        if let (Some((mut current, current_path)), Some((mut neighbor, neighbor_path))) = (
            self.inpaint_image_and_path_from(None)?,
            self.inpaint_image_and_path_from(direction)?,
        ) {
            let (width, height) = (current.width() / 2, current.height());
            let current_subimage = match direction {
                  RIGHT => current.sub_image(width, 0, width, height),
                  BOTTOM => current.sub_image(0, height, width, height),
                  _ => unimplemented!()
           }
           let weighting_function = ... // the same for weighting_function
            interpolate_subimages(
                &mut right.sub_image(0, 0, width, height),
                &mut current_sub_image,
                weighting_function,
            );
            current.save(current_path)?;
            neighbor.save(neighbor_path)?;
        }
        Ok(())
    }
```

Reply:
I would prefer keeping two functions. The generic one would be longer and look like this:
```rust
    fn interpolate_inpaint_image_with_neighbor(&self, direction: Direction) -> ImageResult<()> {
        if let (Some((mut current, current_path)), Some((mut neighbor, neighbor_path))) = (
            self.inpaint_image_and_path_from(None)?,
            self.inpaint_image_and_path_from(direction)?,
        ) {
            let (width, height) = match direction {
                Direction::Right => (current.width() / 2, current.height()),
                Direction::Bottom => (current.width(), current.height() / 2),
                _ => unimplemented!(),
            };
            let (current_subimage, neighbor_subimage) = match direction {
                Direction::Right => (
                    current.sub_image(width, 0, width, height),
                    neighbor.sub_image(0, 0, width, height),
                ),
                Direction::Bottom => (
                    current.sub_image(0, height, width, height),
                    neighbor.sub_image(0, 0, width, height),
                ),
                _ => unimplemented!(),
            };
            let weighting_function = match direction {
                Direction::Right => |i: u32, _: u32| i as f32 / (width - 1) as f32,
                Direction::Bottom => |_: u32, j: u32| j as f32 / (height - 1) as f32,
                _ => unimplemented!(),
            };
            interpolate_subimages(
                &mut neighbor_subimage,
                &mut current_subimage,
                weighting_function,
            );
            current.save(current_path)?;
            neighbor.save(neighbor_path)?;
        }
        Ok(())
    }
```

Reply:
Thought about it a little more this morning. I think I found a quite nice generalization based on yours, so I change it.